### ISSUE TITLE
TimeoutEngine：Add timeout parameter to control actionTask execute time! 

### ISSUE THREAD

**Issue by zhouguoqing** [App-vNext/Polly#514]
```
using System;
using System.Runtime.ExceptionServices;
using System.Threading;
using System.Threading.Tasks;
using Polly.Utilities;

#if NET40
using ExceptionDispatchInfo = Polly.Utilities.ExceptionDispatchInfo;
#endif

namespace Polly.Timeout
{
    internal static partial class TimeoutEngine
    {
        internal static TResult Implementation<TResult>(
            Func<Context, CancellationToken, TResult> action,
            Context context,
            CancellationToken cancellationToken,
            Func<Context, TimeSpan> timeoutProvider,
            TimeoutStrategy timeoutStrategy,
            Action<Context, TimeSpan, Task, Exception> onTimeout)
        {
            cancellationToken.ThrowIfCancellationRequested();
            TimeSpan timeout = timeoutProvider(context);

            using (CancellationTokenSource timeoutCancellationTokenSource = new CancellationTokenSource())
            {
                using (CancellationTokenSource combinedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, timeoutCancellationTokenSource.Token))
                {
                    CancellationToken combinedToken = combinedTokenSource.Token;

                    Task<TResult> actionTask = null;
                    try
                    {
                        if (timeoutStrategy == TimeoutStrategy.Optimistic)
                        {
                            SystemClock.CancelTokenAfter(timeoutCancellationTokenSource, timeout);
                            return action(context, combinedToken);
                        }

                        // else: timeoutStrategy == TimeoutStrategy.Pessimistic

                        SystemClock.CancelTokenAfter(timeoutCancellationTokenSource, timeout);

                        actionTask = 

                        #if NET40
                            TaskEx
                        #else
                            Task
                        #endif                             

                        .Run(() =>
                            action(context, combinedToken)       // cancellation token here allows the user delegate to react to cancellation: possibly clear up; then throw an OperationCanceledException.
                            , combinedToken);           // cancellation token here only allows Task.Run() to not begin the passed delegate at all, if cancellation occurs prior to invoking the delegate.  
                        try
                        {
                           //Add timeout parameter to control actionTask execute time! 
                           var isProcessed =  actionTask.Wait(Convert.ToInt32(timeout.TotalMilliseconds), timeoutCancellationTokenSource.Token);
                            if (isProcessed == false)
                            {                                
                                if (timeoutCancellationTokenSource.IsCancellationRequested == false)
                                    timeoutCancellationTokenSource.Cancel(true);
                                throw new TimeoutRejectedException("The delegate executed through TimeoutPolicy did not complete within the timeout（MS）：" + timeout + ".@" + DateTime.Now.ToString("HH:mm:ss.fff"));
                            }
                            // cancellation token here cancels the Wait() and causes it to throw, but does not cancel actionTask.  
                            // We use only timeoutCancellationTokenSource.Token here, not combinedToken.  
                            // If we allowed the user's cancellation token to cancel the Wait(), 
                            // in this pessimistic scenario where the user delegate may not observe that cancellation, that would create a no-longer-observed task.  
                            // That task could in turn later fault before completing, risking an UnobservedTaskException.
                        }
                        catch (AggregateException ex) when (ex.InnerExceptions.Count == 1) // Issue #270.  Unwrap extra AggregateException caused by the way pessimistic timeout policy for synchronous executions is necessarily constructed.
                        {
                            ExceptionDispatchInfo.Capture(ex.InnerException).Throw();
                        }

                        return actionTask.Result;
                    }
                    catch (TimeoutRejectedException re)
                    {
                        onTimeout(context, timeout, actionTask, re);
                        throw;
                    }
                    catch (Exception ex)
                    {
                        if (timeoutCancellationTokenSource.IsCancellationRequested)
                        {
                            onTimeout(context, timeout, actionTask, ex);
                            throw new TimeoutRejectedException("The delegate executed through TimeoutPolicy did not complete within the timeout.", ex);
                        }

                        throw;
                    }
                }
            }
        }

    }
}
```

(**EDITED**: by @reisenberger for layout only)

**Comment by reisenberger** [comment]
Thank you @zhouguoqing for sharing this and for joining the Polly conversation.

**TL;DR**:

+ The thread pool behaviour which the StackOverflow question demonstrates is known behaviour of the .NET threadpool and not connected to Polly TimeoutPolicy.
+ The [change proposed to `TimeoutEngine`](https://github.com/App-vNext/Polly/issues/514#issue-368206967) makes no difference to the way `actionTask` executes: it does not control `actionTask` execute time.

---

### Threadpool behaviour

+ The thread pool behaviour which the StackOverflow question demonstrates is known behaviour of the .NET threadpool and not specifically connected to Polly TimeoutPolicy. If you take TimeoutPolicy out of the code in the StackOverflow question, you should still see thread starvation.
+ [By default, the initial value of ThreadPool MinThreads is set to the number of cores on the system](https://docs.microsoft.com/en-us/dotnet/api/system.threading.threadpool.getminthreads?view=netframework-4.7.2). This probably explains the rapid execution seen for the first _8_ parallel operations in the StackOverflow question.  After that, the default .NET ThreadPool implementation injects one new thread per half second (500ms), or more frequently if other threads become free more quickly. (That is a brief explanation. Matt Warren has blogged [this good explanation of the Thread Pool's algorithms](http://mattwarren.org/2017/04/13/The-CLR-Thread-Pool-Thread-Injection-Algorithm/) and some [Microsoft documentation is here](https://docs.microsoft.com/en-us/previous-versions/msp-n-p/ff963549(v=pandp.10)#thread-injection).)
+ Yes, [`ThreadPool.SetMinThreads()`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.threadpool.setminthreads) is one way to force the ThreadPool to creating a greater number of threads initially (notwithstanding the Microsoft [caveats](https://docs.microsoft.com/en-us/dotnet/api/system.threading.threadpool.setminthreads?view=netframework-4.7.2#remarks)), if you consciously want a greater number of operations to start in parallel without waiting for ThreadPool injection to inject more threads.
+ Polly's TimeoutPolicy does not interact with the ThreadPool, except that `TimeoutStrategy.Pessimistic` on synchronous executions consumes an extra ThreadPool thread, [as we explicitly document](https://github.com/App-vNext/Polly/wiki/Timeout#additional-notes-pessimistic-timeout-for-sync-executions).

---

### The proposed change

The [change proposed to `TimeoutEngine`](https://github.com/App-vNext/Polly/issues/514#issue-368206967) proposes:

    //Add timeout parameter to control actionTask execute time!
    actionTask.Wait(Convert.ToInt32(timeout.TotalMilliseconds), timeoutCancellationTokenSource.Token);

instead of:

    actionTask.Wait(timeoutCancellationTokenSource.Token);

This change has no effect on the way `actionTask` executes: it does not control `actionTask` execute time.  This can be verified with the following code:

    public static void Main()
    {
        const int sleepMs = 200;
        const int timeoutMs = 10;

        Action work = () => Thread.Sleep(sleepMs);
        long before = Stopwatch.GetTimestamp();
        Task showWhatHappensToWork = Task.Run(work).ContinueWith(t =>
        {
            long elapsedUntilWorkCompleted = (Stopwatch.GetTimestamp() - before) /  TimeSpan.TicksPerMillisecond;
            Console.WriteLine("After " + elapsedUntilWorkCompleted.ToString("D3") + " ms: Work " + (t.IsFaulted ? "faulted" : (t.IsCanceled ? "canceled" : "completed (was not canceled)")));
        }
        );

        showWhatHappensToWork.Wait(timeoutMs);

        long elapsedUntilStoppedWaiting = (Stopwatch.GetTimestamp() - before) /  TimeSpan.TicksPerMillisecond;
        Console.WriteLine("After " + elapsedUntilStoppedWaiting.ToString("D3") + " ms: Stopped waiting (walked away from waiting)." );

        showWhatHappensToWork.Wait(); // To allow work enough time to complete, and Console.Writeline() its output.
    }

This code can be tried [in dotnetfiddle here](https://dotnetfiddle.net/zQk3f9).  Note that `.Wait(timeoutMs)` does not make `work` terminate any sooner. It still runs to `completed (was not canceled)` at around 200ms.  This is the behaviour we discuss in TimeoutPolicy documentation under [What happens to the timed-out delegate?](https://github.com/App-vNext/Polly/wiki/Timeout#what-happens-to-the-timed-out-delegate).

All of the available `Task.Wait()` overloads merely control how long the calling thread will _wait_ for the `Task` instance to complete, after which time the calling thread _walks away from waiting_ (i.e. stops waiting), as we [document for TimeoutPolicy](https://github.com/App-vNext/Polly/wiki/Timeout#pessimistic-timeout). None of the `Task.Wait()` overloads terminate the work represented by the `Task` instance being `.Wait()`-ed.

`actionTask.Wait(Convert.ToInt32(timeout.TotalMilliseconds), timeoutCancellationTokenSource.Token)` will just provide two different timing mechanisms, and these two timing mechanisms will race each other to trigger the walking-away-from-waiting,

---

Hope that helps, but please let me know if I have misunderstood the intention of [the suggestion](https://github.com/App-vNext/Polly/issues/514#issue-368206967).
